### PR TITLE
Add onResult to AvailableIntervention.

### PR DESCRIPTION
**Pull request by mhkawano**
Adds `onResult` support in `AvailableIntervention`. This replaces `setOnInterventionComplete` which was originally planned to be used for "Funnel Prompts for Enterprise". This was chosen to simplify the API for third parties.

### CODE REVIEW

File: src/runtime/audience-action-flow.js [hunk 1]
```diff
@@ -290,11 +291,14 @@ export class AudienceActionFlow {
...
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  handleSurveyDataTransferRequest_(request) {
+  async handleSurveyDataTransferRequest_(request) {
     // @TODO(justinchou): execute callback with setOnInterventionComplete
     // then check for success
-    const gaLoggingSuccess = this.logSurveyDataToGoogleAnalytics(request);
-    if (gaLoggingSuccess) {
+    const {onResult} = this.params_;
+    const loggingSuccess = onResult
+      ? await this.params_.onResult(request)
```
Comment (ChrisAntaki) [inline src/runtime/audience-action-flow.js:299]:
```suggestion
      ? await onResult(request)
```

Comment (ChrisAntaki) [reply src/runtime/audience-action-flow.js:299]: We might want to gracefully support rejects/throws from the `onResult` callback

Comment (ChrisAntaki) [reply src/runtime/audience-action-flow.js:299]: We could `await onResult(request)` within a `try/catch` block, and if it fails, we could set `loggingSuccess` to false

File: src/runtime/audience-action-flow.js [hunk 2]
```diff
@@ -290,11 +291,14 @@ export class AudienceActionFlow {
...
    */
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  handleSurveyDataTransferRequest_(request) {
+  async handleSurveyDataTransferRequest_(request) {
     // @TODO(justinchou): execute callback with setOnInterventionComplete
     // then check for success
-    const gaLoggingSuccess = this.logSurveyDataToGoogleAnalytics(request);
-    if (gaLoggingSuccess) {
+    const {onResult} = this.params_;
+    const loggingSuccess = onResult
```
Comment (oyj9109) [inline src/runtime/audience-action-flow.js:298]: nit: what about 'dataTransferSuccess'?

File: src/runtime/audience-action-flow-test.js
Comment (oyj9109) [inline src/runtime/audience-action-flow-test.js:1]: nit: can you rename the existing test cases to be specfiic to Google Analytics?

File: src/runtime/audience-action-flow.js (hunk 1 above)
Comment (oyj9109) [reply src/runtime/audience-action-flow.js:299]: Good catch! +1 to error handling

**Review by oyj9109 — APPROVED** [review]

File: src/runtime/audience-action-flow.js (hunk 1 above)
Comment (mhkawano) [reply src/runtime/audience-action-flow.js]: Moved this area into a new function `attemptSurveyDataTransfer` and wrapped `onResult` in a try/catch with a warning if it catches an exception.

File: src/runtime/audience-action-flow.js (hunk 2 above)
Comment (mhkawano) [reply src/runtime/audience-action-flow.js]: Done.

File: src/runtime/audience-action-flow-test.js
Comment (mhkawano) [reply src/runtime/audience-action-flow-test.js:1]: Done.

**Comment by mhkawano** [comment]
>BTW, setOnInterventionComplete is not limited to RRM:E. We will need to provide an API for RRM:S to be able to trasnfer data to their system. I agree getting a callback onResult as a parameter of show function for RRM:E would be more convenient for them, but are we okay with supporting setOnInterventionComplete later for RRM:S and maintaining two different paths for setting callbacks?
I didn't mean to imply that it would replace the use cases for RRM:S. The consensus in the after review discussion seemed to be that we were okay with two paths.

File: src/runtime/audience-action-flow.js [hunk 3]
```diff
@@ -311,10 +311,31 @@ export class AudienceActionFlow {
...
+   * @return {boolean}
+   * @private
+   */
+  async attemptSurveyDataTransfer(request) {
+    // @TODO(justinchou): execute callback with setOnInterventionComplete
+    // then check for success
+    const {onResult} = this.params_;
+    if (onResult) {
+      try {
+        return onResult(request);
```
Comment (ChrisAntaki) [inline src/runtime/audience-action-flow.js:330]: I think this should `await` the onResult call, in case it returns a promise

Comment (ChrisAntaki) [reply src/runtime/audience-action-flow.js:330]: It could return a promise that eventually rejects

Comment (mhkawano) [reply src/runtime/audience-action-flow.js]: Done.

File: src/runtime/audience-action-flow-test.js [hunk 4]
```diff
@@ -713,7 +722,147 @@ describes.realWin('AudienceActionFlow', (env) => {
...
+    const messageCallback = messageMap[TEST_SURVEYDATATRANSFERREQUEST.label()];
+    messageCallback(TEST_SURVEYDATATRANSFERREQUEST);
+
+    await tick(10);
+
+    activityIframeViewMock.verify();
+    onResultMock.verify();
+  });
+
+  it(`handles a SurveyDataTransferRequest with onResult logging exception`, async () => {
```
[the remark this answers is not in the document]
Comment (mhkawano) [reply src/runtime/audience-action-flow-test.js:819]: Done.

File: src/runtime/audience-action-flow-test.js [hunk 5]
```diff
@@ -713,7 +722,196 @@ describes.realWin('AudienceActionFlow', (env) => {
...
+    );
+    activityIframeViewMock.verify();
+    onResultMock.verify();
+  });
+
+  it(`handles a SurveyDataTransferRequest with onResult logging rejection`, async () => {
+    const onResultMock = sandbox
+      .mock()
+      .withExactArgs(TEST_SURVEYDATATRANSFERREQUEST)
+      .resolves(Promise.reject(new Error('Test Callback Exception')))
```
Comment (ChrisAntaki) [inline src/runtime/audience-action-flow-test.js:872]: Would this work?
```suggestion
      .rejects(new Error('Test Callback Exception'))
```

**Review by ChrisAntaki — APPROVED** [review]
Looks good 👍

**Pull request merged**